daemon/top_unix.go

```diff
@@ -58,25 +75,38 @@ func parsePSOutput(output []byte, pids []int) (*container.ContainerTopOKBody, er
 	}
 
 	// loop through the output and extract the PID from each line
+	// fixing #30580, be able to display thread line also when "m" option used
+	// in "docker top" client command
+	preContainedPidFlag := false
 	for _, line := range lines[1:] {
 		if len(line) == 0 {
 			continue
 		}
 		fields := fieldsASCII(line)
-		p, err := strconv.Atoi(fields[pidIndex])
-		if err != nil {
-			return nil, fmt.Errorf("Unexpected pid '%s': %s", fields[pidIndex], err)
-		}
 
-		for _, pid := range pids {
-			if pid == p {
-				// Make sure number of fields equals number of header titles
-				// merging "overhanging" fields
-				process := fields[:len(procList.Titles)-1]
-				process = append(process, strings.Join(fields[len(procList.Titles)-1:], " "))
-				procList.Processes = append(procList.Processes, process)
+		var (
+			p   int
+			err error
+		)
+
+		if fields[pidIndex] == "-" {
+			if preContainedPidFlag {
+				appendProcess2ProcList(procList, fields)
+			}
+			continue
+		} else {
```

Comment:
Maybe I'm not quite clear get you.  Let me clear it a little by below example except from my test:
{
    {[]byte(`  PID COMMAND
   42 foo
   43 bar
    - -
    - -
   41 foo
   50 kk
    - -
  100 baz
    - -
`), []int{50, 43, 100}, false}:

According to the logic for "docker top" command, only the processes {50,43,100} should be displayed (which are the container contained processes, others not belong to this container, so ignore them):
"42 foo", not in the pids, so ignore,  preContainedPidFlag = false
"43 bar", should display.  preContainedPidFlag = true
 "- -"  , following above,  preContainedPidFlag = true
 "- -" ,  should display,  preContainedPidFlag = true //continer, else it lead to  preContainedPidFlag = true at the bottom 
"41 foo", not in the pids, so ignore,  preContainedPidFlag = false
"100 baz", should display, preContainedPidFlag = true
"- -", should display, preContainedPidFlag = true

So it works.  

Yes, this can be refactored into below, do you intend to have code this way (much clear :) ), or let it be, thanks. 
if thing {
  // do stuff
  continue
}
// do other stuff